### PR TITLE
fix(curriculum): add a lesson to registration form practice project for requiring radio buttons

### LINKED ISSUE #47456: Learn HTML Forms by Building a Registration Form
### Is your feature request related to a problem? Please describe.

The registration form currently accepts submissions with both radio buttons unchecked. The Account-Type must be a required parameter otherwise it beats the purpose of having it in the first place.

### Describe the solution you'd like

The simplest solution would be to add a checked attribute to any one of the two radio buttons.

### Describe alternatives you've considered

Or leave them both unchecked initially, but force the user to check any one of them when they try to submit the form (which I don't know how to do)

### Additional context

Here's the link: https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-html-forms-by-building-a-registration-form/step-41

### PR DESCRIPTION
**Pull request by NiharPhansalkar**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #47456

User can make it mandatory to select a radio button using the `required` attribute.
cypressTests

### CODE REVIEW

**Comment by NiharPhansalkar** [comment]
The request was to add a `checked` attribute to the buttons and create a legend. However, I personally feel that a `required` attribute is the best and the cleanest way to address this issue. This leaves both radio buttons unchecked initially, however, does not allow any user to submit the form without choosing one. And since all radio buttons are grouped together with the `name` attribute, the `required` attribute is applied on the entire group of buttons, so adding new buttons will maintain the `required` attribute provided they have the same `name` attribute to them.

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+To correct this add a `required` attribute to `personal-account or business-account`.
+
+# --hints--
+
+You should use the `required` attribute for the `input` tag.
+
+```js
+assert.equal(document.getElementsByName("account-type").required, true);
+```
```
Comment (ieahleen) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:22]: this test will fail, you need to update the seed code of all following steps

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      <input type="submit" value="Submit" />
+    </form>
+  </body>
+</html>
+```
+
+### Code Explanation
+
+Since our radio buttons are linked together due to the name attribute, we can give a required field to any of the radio buttons, and it will apply to the entire group.
+
```
Comment (ieahleen) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:80]: this can't go in here. Explanations can go only under the description

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    <p>Please fill out this form with the required information</p>
+    <form action='https://register-demo.freecodecamp.org'>
+      <fieldset>
+        <label for="first-name">Enter Your First Name: <input id="first-name" name="first-name" type="text" required /></label>
+        <label for="last-name">Enter Your Last Name: <input id="last-name" name="last-name" type="text" required /></label>
+        <label for="email">Enter Your Email: <input id="email" name="email" type="email" required /></label>
+        <label for="new-password">Create a New Password: <input id="new-password" name="new-password" type="password" pattern="[a-z0-5]{8,}" required /></label>
+      </fieldset>
+      <fieldset>
+        <label for="personal-account"><input id="personal-account" type="radio" name="account-type" required/> Personal Account</label>
```
Comment (ieahleen) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:49]: Let the user add it, do not add it yourself

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]
The project needs to have the required attribute in all the following steps

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md (hunk 1 above)
Comment (NiharPhansalkar) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:22]: So I should add this test to all subsequent steps and remove it from this one?

Comment (ieahleen) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:22]: No, the test stay here. But you need to update the seed code of all the following steps and the final solution

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md (hunk 2 above)
Comment (NiharPhansalkar) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md]: Screenshot from 2022-10-02 23-28-35
Is there an issue open to update this?

Comment (NiharPhansalkar) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md]: https://contribute.freecodecamp.org/#/how-to-work-on-coding-challenges

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/60ffefd6479a3d084fb77cbc.md [hunk 4]
```diff
@@ -48,8 +48,8 @@ assert.equal(new __helpers.CSSHelp(document).getStyle('a')?.color, 'rgb(223, 223
         <label for="new-password">Create a New Password: <input id="new-password" name="new-password" type="password" pattern="[a-z0-5]{8,}" required /></label>
       </fieldset>
       <fieldset>
-        <label for="personal-account"><input id="personal-account" type="radio" name="account-type" class="inline" /> Personal Account</label>
```
Comment (ieahleen) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/60ffefd6479a3d084fb77cbc.md:51]: why did you delete `class="inline"`?

**Comment by ieahleen** [comment]
About that screenshot...
The correct test to run is
`npm run test:curriculum --block='Learn HTML Forms by Building a Registration Form'`

Your changes will fail. You have removed code added in some of the later stepts

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    <form action='https://register-demo.freecodecamp.org'>
+      <fieldset>
+        <label for="first-name">Enter Your First Name: <input id="first-name" name="first-name" type="text" required /></label>
+        <label for="last-name">Enter Your Last Name: <input id="last-name" name="last-name" type="text" required /></label>
+        <label for="email">Enter Your Email: <input id="email" name="email" type="email" required /></label>
+        <label for="new-password">Create a New Password: <input id="new-password" name="new-password" type="password" pattern="[a-z0-5]{8,}" required /></label>
+      </fieldset>
+      <fieldset>
+        <label for="personal-account"><input id="personal-account" type="radio" name="account-type" /> Personal</label>
+        <label for="business-account"><input id="business-account" type="radio" name="account-type" /> Business</label>
```
Comment (ieahleen) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:52]: You never added the `legend`, so now it doesn't make sense to have `Personal` and `Business` on their own there

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+You have created two radio buttons. However, if you submit the form, it does not require you to check any of them.
+
+This is not good, since you need to know what type of account the user requires.
+
+To correct this add a `required` attribute to `personal-account or business-account`.
+
+# Code Explanation
+
+Since our `radio buttons` are linked together due to the `name` attribute, we can give the `required` field to any of the radio buttons, and it will apply to the entire group.
```
Comment (ieahleen) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:18]: Don't add an extra header, follow the template

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/60ffefd6479a3d084fb77cbc.md [hunk 7]
```diff
@@ -48,8 +48,8 @@ assert.equal(new __helpers.CSSHelp(document).getStyle('a')?.color, 'rgb(223, 223
...
-        <label for="business-account"><input id="business-account" type="radio" name="account-type" class="inline" /> Business Account</label>
+        <label for="personal-account"><input id="personal-account" type="radio" name="account-type" required/> Personal</label>
+        <label for="business-account"><input id="business-account" type="radio" name="account-type" /> Business</label>
```
Comment (ieahleen) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/60ffefd6479a3d084fb77cbc.md:52]: do not delete code added in future steps. That `class="inline"` must stay

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]

**Comment by NiharPhansalkar** [comment]
I am extremely sorry for the mistakes. Thank you for being patient with me.

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]
you are still missing the legend element. Please check again the issue.

**Comment by bbsmooth** [comment]
@NiharPhansalkar 
> The request was to add a checked attribute to the buttons and create a legend. However, I personally feel that a required attribute is the best and the cleanest way to address this issue.

And you have checked your implementation with a screen reader to make sure it functions appropriately? There is a reason we settled on that particular solution in the issue. If you feel that your solution is better and can justify it then please make the argument in the issue. But until the consensus on the solution has changed in the issue then I don't think you should be able to make the decision by yourself that it should be implemented differently.

I'm afraid I have to recommend that this PR not be approved because it does not follow accessibility best practices for implementing a required radio button grouping.

**Comment by NiharPhansalkar** [comment]
@bbsmooth I have made the required changes. I changed all `required` into `checked` and also added in the legend. Please re-review the PR.

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    <form action='https://register-demo.freecodecamp.org'>
+      <fieldset>
+        <label for="first-name">Enter Your First Name: <input id="first-name" name="first-name" type="text" required /></label>
+        <label for="last-name">Enter Your Last Name: <input id="last-name" name="last-name" type="text" required /></label>
+        <label for="email">Enter Your Email: <input id="email" name="email" type="email" required /></label>
+        <label for="new-password">Create a New Password: <input id="new-password" name="new-password" type="password" pattern="[a-z0-5]{8,}" required /></label>
+      </fieldset>
+      <fieldset>
+        <label for="personal-account"><input id="personal-account" type="radio" name="account-type" /> Personal Account</label>
+        <label for="business-account"><input id="business-account" type="radio" name="account-type" /> Business Account</label>
```
Comment (ieahleen) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:52]: Shouldn't these be just 'Personal' and 'Business'?

These are added in a previous step, so probably that needs changing too

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+Remove the `Account` in the `label` tag. The `legend` element will let the user know what is `Personal` and `Business`.
+
+# --hints--
+
+You should use the `checked` attribute for the `input` tag.
+
+```js
+assert.equal(document.getElementsByName("account-type").required, true);
+```
```
Comment (ieahleen) [inline curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:26]: this still checks for required, not checked.

There is no tests for the `legend` element

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md (hunk 8 above)
Comment (NiharPhansalkar) [reply curriculum/challenges/english/14-responsive-web-design-22/learn-html-forms-by-building-a-registration-form/633928c6b0baf16c0b0d823b.md:52]: The user has not added a `legend` at this point, so I left the `Account` there to not break the flow of the previous lectures. If this changes, then I will have to change all the previous steps too right?

**Comment by raisedadead** [comment]
@bbsmooth @ieahleen 

Please review, but we will try and land this later than usual because it adds files which is a logistical problem - for deployments.

We will try and fit this into the queue next week. But that should not stop you from reviewing it.

Thanks.

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]
please update the solution in the last step

image

**Comment by NiharPhansalkar** [comment]
Sure, I will look into it over the weekend and make the required changes. 👍

**Comment by Sboonny** [comment]
@NiharPhansalkar, how is this going so far?
